bootstore/src/schemes/v0/peer.rs

```diff
@@ -0,0 +1,797 @@
+                initial_membership,
+                responder: tx,
+            })
+            .await?;
+        let res = rx.await?;
+        res
+    }
+
+    /// Initialize this peer as a learner
+    pub async fn init_learner(&self) -> Result<(), PeerRequestError> {
+        let (tx, rx) = oneshot::channel();
+        self.tx.send(PeerApiRequest::InitLearner { responder: tx }).await?;
+        let res = rx.await?;
+        res
+    }
+
+    /// Load the rack secret.
+    ///
+    /// This can only be successfully called when a peer has been initialized,
+    /// either as initial member or learner who has learned its share.
+    pub async fn load_rack_secret(
+        &self,
+    ) -> Result<RackSecret, PeerRequestError> {
+        let (tx, rx) = oneshot::channel();
+        self.tx.send(PeerApiRequest::LoadRackSecret { responder: tx }).await?;
+        let res = rx.await?;
+        res
+    }
+
+    /// Inform the peer of currently known IP addresses on the bootstrap network
+    ///
+    /// These are generated from DDM prefixes learned by the bootstrap agent.
+    pub async fn load_peer_addresses(
+        &self,
+        addrs: BTreeSet<SocketAddrV6>,
+    ) -> Result<(), PeerRequestError> {
+        self.tx.send(PeerApiRequest::PeerAddresses(addrs)).await?;
+        Ok(())
+    }
+
+    /// Get the status of this peer
+    pub async fn get_status(&self) -> Result<Status, PeerRequestError> {
+        let (tx, rx) = oneshot::channel();
+        self.tx.send(PeerApiRequest::GetStatus { responder: tx }).await?;
+        let res = rx.await?;
+        Ok(res)
+    }
+
+    /// Shutdown the peer
+    pub async fn shutdown(&self) -> Result<(), PeerRequestError> {
+        self.tx.send(PeerApiRequest::Shutdown).await?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Status {
+    fsm_state: &'static str,
+    peers: BTreeSet<SocketAddrV6>,
+    connections: BTreeMap<Baseboard, SocketAddrV6>,
+    accepted_connections: BTreeSet<SocketAddrV6>,
+    negotiating_connections: BTreeSet<SocketAddrV6>,
+}
+
+/// A peer in the bootstore protocol
+pub struct Peer {
+    config: Config,
+    fsm: Fsm,
+    peers: BTreeSet<SocketAddrV6>,
+    handle_unique_id_counter: u64,
+
+    // boolean set when a `PeerApiRequest::shutdown` is received
+    shutdown: bool,
+
+    // Connections that have been accepted, but where negotiation has not
+    // finished At this point, we only know the client port of the connection,
+    // and so cannot identify  it as a `Peer`.
+    accepted_connections: BTreeMap<SocketAddrV6, AcceptedConnHandle>,
+
+    // Connections that have not yet completed handshakes via `Hello` and
+    // `Identify` messages.
+    //
+    // We only store the client (connecting) side, not server (accepting) side here.
+    negotiating_connections: BTreeMap<SocketAddrV6, PeerConnHandle>,
+
+    // Active connections participating in scheme v0
+    //
+    // This consists of both client and server connections
```

Comment:
The "client and server" terminology confuses me here.  It seems like connections flow through `accepted_connections` -> `negotiating_connections` -> `connections`, but the first two only store client connections and the latter somehow also has server connections?

I'll keep reading...

Reply:
Honestly, me too. What I'm really trying to get at is that one side is the connection initiating side, and one side accepts connections. We have to manage them separately at the beginning, but once the handshake is complete, they are treated equivalently. I've written this type of code in multiple projects in the past and always struggled naming these things. 

Maybe `initiator` and `acceptor` ?

Reply:
fixed in c6735245f